mkosi/kmod.py

```diff
@@ -349,14 +349,56 @@ def gen_required_kernel_modules(
     # Include or exclude firmware explicitly configured
     firmware = filter_firmware(context.root, firmware, include=firmware_include, exclude=firmware_exclude)
 
-    # Some firmware dependencies are symbolic links, so the targets for those must be included in the list
-    # of required firmware files too. Intermediate symlinks are not included, and so links pointing to links
-    # results in dangling symlinks in the final image.
-    for fw in firmware.copy():
-        if (context.root / fw).is_symlink():
-            target = Path(chase(os.fspath(context.root), os.fspath(fw)))
-            if target.exists():
-                firmware.add(target.relative_to(context.root))
+    # /usr/lib/firmware makes use of symbolic links so we have to make sure the symlinks and their targets
+    # are all included.
+    todo = firmware.copy()
+    firmware.clear()
+    for fw in todo:
+        # Every path component from /usr/lib/firmware up to and including the firmware file itself might be a
+        # symlink. We need to make sure we include all of them so we iterate over them and keep resolving
+        # each symlink separately (and recursively) and add all of them to the list of firmware to add.
+        #
+        # As of the time of writing this logic, the only firmware that actually requires intermediate path
+        # symlink resolution are the following:
+        #
+        # $ find /usr/lib/firmware -type l | grep -v "\."
+        # /usr/lib/firmware/intel/sof-ace-tplg
+        # /usr/lib/firmware/nvidia/ad103
+        # /usr/lib/firmware/nvidia/ad104
+        # /usr/lib/firmware/nvidia/ad106
+        # /usr/lib/firmware/nvidia/ad107
+        # /usr/lib/firmware/nvidia/ga103/gsp
+        # /usr/lib/firmware/nvidia/ga104/gsp
+        # /usr/lib/firmware/nvidia/ga106/gsp
+        # /usr/lib/firmware/nvidia/ga107/gsp
+        # /usr/lib/firmware/nvidia/gb102
+        # /usr/lib/firmware/nvidia/gb203
+        # /usr/lib/firmware/nvidia/gb205
+        # /usr/lib/firmware/nvidia/gb206
+        # /usr/lib/firmware/nvidia/gb207
+        # /usr/lib/firmware/nvidia/tu104/gsp
+        # /usr/lib/firmware/nvidia/tu106/gsp
+        # /usr/lib/firmware/nvidia/tu117/gsp
+
+        for p in (*reversed(parents_below(context.root / fw, context.root / firmwared)), context.root / fw):
```

Comment:
The unpacking makes this more difficult to understand, I think it would be easier to add a functio
```python
def path_and_parents_below(path: Path, below: Path) -> list[Path]:
    return [path] + parents_below(path, below)
```
and use that here